convolve.c

```diff
@@ -0,0 +1,151 @@
+ * @param imageNumPix number of columns x number of rows in image
+ * @param[out] neighborhood 9-element array containing 3x3 block of pixels around
+ * `pixelIndex`
+ */
+void getNeighborhood(
+    uint16_t* imageBuffer,
+    uint16_t imageMean,
+    uint32_t pixelIndex,
+    uint16_t imageWidth,
+    uint32_t imageNumPix,
+    uint16_t* neighborhood)
+{
+    bool isLeft = (0 == pixelIndex % imageWidth);
+    bool isRight = (0 == ((pixelIndex + 1) % imageWidth));
+    bool isBottom = ((int64_t)pixelIndex - imageWidth) < 0;
+    bool isTop = (pixelIndex + imageWidth) >= imageNumPix;
+
+    // HACK: FIXME: sobelmetric returns to 0 when high?
+    imageMean = 0U;
+
+    // If not blocked by image bound guards, return pixel value, else return image mean value
+    neighborhood[0] = (!isBottom && !isLeft)     ? imageBuffer[pixelIndex - imageWidth - 1] : imageMean;
+    neighborhood[1] = (!isBottom)                ? imageBuffer[pixelIndex - imageWidth]     : imageMean;
+    neighborhood[2] = (!isBottom && !isRight)    ? imageBuffer[pixelIndex - imageWidth + 1] : imageMean;
+    neighborhood[3] = (!isLeft)                  ? imageBuffer[pixelIndex - 1]              : imageMean;
+    // ought to be limited by caller's for loop, but belt & suspenders
+    neighborhood[4] = (pixelIndex < imageNumPix) ? imageBuffer[pixelIndex]                  : imageMean;
+    neighborhood[5] = (!isRight)                 ? imageBuffer[pixelIndex + 1]              : imageMean;
+    neighborhood[6] = (!isTop && !isLeft)        ? imageBuffer[pixelIndex + imageWidth - 1] : imageMean;
+    neighborhood[7] = (!isTop)                   ? imageBuffer[pixelIndex + imageWidth]     : imageMean;
+    neighborhood[8] = (!isTop && !isRight)       ? imageBuffer[pixelIndex + imageWidth + 1] : imageMean;
+}
+
+
+/**
+ * @brief The innermost part of the convolution algorithm, add-and-multiply
+ * pixels by kernel elements.
+ * 
+ * @param[in] array the input array to be considered; intended to be the output of
+ * `getNeighborhood`
+ * @param kernel the convolution kernel. Don't pass arrays of other types in
+ * here, you heathen.
+ * @param kernelSize the length of the flattened, square kernel
+ * @return the result of the convolution
+ */
+float convolve(uint16_t* array, int16_t* kernel, uint8_t kernelSize)
+{
+    float sum = 0;
+    // Assumption: array and kernel are same size.
+    for (uint8_t i = 0; i < kernelSize; i++) {
+        sum += (float)array[i] * (float)kernel[kernelSize - i - 1];
+    }
+    return sum;
+}
+
+
+/**
+ * @brief The innermost part of the convolution algorithm, add-and-multiply
+ * pixels by kernel elements. Optimized for convolving 2 3x3 arrays
+ * 
+ * @param[in] array the input array to be considered; intended to be the output of
+ * `getNeighborhood`
+ * @param kernel the convolution kernel. Don't pass arrays of other types in
+ * here, you heathen.
+ * @return the result of the convolution
+ */
+float convolve9(uint16_t* array, int16_t* kernel)
```

Comment:
I still have questions about the kernel used here, is it the fractional kernel or the integer kernel that gets divided out by later. Not serious enough to be an issue just curious

Reply:
The Sobel kernel is integer-valued. A Gaussian kernel could be integer-valued, as long as you remember to divide the entire image by, e.g. 16 afterward. This is just the integer kernel part of that